### PR TITLE
refactor(model): refactor model architecture

### PR DESCRIPTION
**Pull request by mwiacx**
## Design

整体设计：成为中间统一的接口层
image

## TODO
1. 初稿
- [x] module
    - [x] linear
    - [x] attention
    - [x] mlp
    - [x] embedding
    - [x] layernorm
    - [x] coustom support
- [x] flash-attn and non-flash-attn
- [x] pack and non-pack 
- [x] support pp
2. 润色
- [x] 校对和添加注释
- [x] 英文拼写校对

### CODE REVIEW

File: internlm/core/parallel/comm/isp.py [hunk 1]
```diff
@@ -573,3 +588,164 @@ def after_backward(self, scheduler, inputs_grad) -> None:
...
+class DistributedAttention(nn.Module):
+    """Initialization.
+
+    Arguments:
+        local_attention (Module): local attention with q,k,v
+        sequence_process_group (ProcessGroup): sequence parallel process group
+        first_scatter_idx (int): scatter_idx for the first all2all comm
+        first_gather_idx (int): gather_idx for the first all2all comm
+        second_scatter_idx (int): scatter_idx for the second all2all comm
+        second_gather_idx (int): gather_idx for the second all2all comm
```
Comment (huangting4201) [inline internlm/core/parallel/comm/isp.py]: 注释的参数列表可能要更新一下

File: internlm/core/parallel/comm/isp.py [hunk 2]
```diff
@@ -573,3 +588,164 @@ def after_backward(self, scheduler, inputs_grad) -> None:
...
+            kwargs: other args
+
+        Returns:
+            * output (Tensor): context output
+        """
+        # qkv shape: [1, packlen, 3, n_head, head_dim] or [batch, seqlen, 3, n_head, head_dim]
+        # scatter in n_head and gather in seqlen(packlen)
+        qkv = _SeqAllToAll.apply(self.spg, qkv, 3, 1)
+
+        context = self.local_attn(qkv, **kwargs)
```
Comment (huangting4201) [inline internlm/core/parallel/comm/isp.py:694]: 这里调用local_attn时，传参要指明参数名字，如self.local_attn(qkv=qkv, **kwargs)，不然会有bug，参考pr https://github.com/InternLM/InternEvo/pull/172

File: internlm/core/parallel/comm/isp.py [hunk 3]
```diff
@@ -573,3 +588,164 @@ def after_backward(self, scheduler, inputs_grad) -> None:
...
+            output (Tensor): context output
+        """
+        # q shpae: [1, packlen, n_head, head_dim] or [batch, seqlen, n_head, head_dim]
+        # scatter in n_head and gather in seqlen(packlen)
+        q = _SeqAllToAll.apply(self.spg, q, 2, 1)
+        # kv shape: [1, packlen, 2, n_head, head_dim] or [batch, seqlen, 2, n_head, head_dim]
+        # scatter in n_head and gather in seqlen(packlen)
+        kv = _SeqAllToAll.apply(self.spg, kv, 3, 1)
+
+        context = self.local_attn(q, kv, **kwargs)
```
Comment (huangting4201) [inline internlm/core/parallel/comm/isp.py:722]: 这里同上

File: internlm/core/parallel/comm/isp.py [hunk 4]
```diff
@@ -573,3 +588,164 @@ def after_backward(self, scheduler, inputs_grad) -> None:
...
+        # scatter in n_head and gather in seqlen(packlen)
+        q = _SeqAllToAll.apply(self.spg, q, 2, 1)
+        # k shpae: [1, packlen, n_head, head_dim] or [batch, seqlen, n_head, head_dim]
+        # scatter in n_head and gather in seqlen(packlen)
+        k = _SeqAllToAll.apply(self.spg, k, 2, 1)
+        # v shpae: [1, packlen, n_head, head_dim] or [batch, seqlen, n_head, head_dim]
+        # scatter in n_head and gather in seqlen(packlen)
+        v = _SeqAllToAll.apply(self.spg, v, 2, 1)
+
+        context = self.local_attn(q, k, v, **kwargs)
```
Comment (huangting4201) [inline internlm/core/parallel/comm/isp.py:755]: 这里也是

File: internlm/core/parallel/comm/utils.py [hunk 5]
```diff
@@ -0,0 +1,229 @@
...
+
+    @staticmethod
+    def backward(ctx, grad_output: Tensor):
+        gather_dim = ctx.reduce_dim
+        grad_input, _ = all_gather_raw(grad_output, ctx.process_group, gather_dim=gather_dim)
+        return grad_input, None, None
+
+
+# Supports autograd, but does not support async
+reduce_scatter = ReduceScatterFunc.apply
```
Comment (huangting4201) [inline internlm/core/parallel/comm/utils.py:58]: ReduceScatterFunc这块要跟develop分支同步下，支持传入reduce_dim

File: internlm/core/parallel/comm/utils.py [hunk 6]
```diff
@@ -0,0 +1,229 @@
...
+        output = memory_pool_allocator(tuple(shape_list))
+    else:
+        output = torch.empty(
+            shape_list,
+            dtype=input_.dtype,
+            device=input_.device,
+        ).contiguous()
+
+    handle = dist.reduce_scatter_tensor(output, input_.contiguous(), op=op, group=process_group, async_op=async_op)
+    return output, handle
```
Comment (huangting4201) [inline internlm/core/parallel/comm/utils.py:226]: reduce_scatter_raw这个函数要跟develop分支同步下

File: internlm/core/parallel/shard.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    if linear_name in ("head", "output"):
+        return "head"
+    elif linear_name in ("wqkv", "wq", "wk", "wv", "wkv", "w1", "w3", "w13"):
+        return "column"
+    elif linear_name in ("wo", "out_proj", "w2") and tp_mode == "isp":
+        return "column"
+    elif linear_name in ("wo", "out_proj", "w2"):
+        return "row"
+    else:
+        return "unknown"
```
Comment (huangting4201) [inline internlm/core/parallel/shard.py:44]: 这里是否要raise error退出更安全一些呢

File: internlm/core/scheduler/no_pipeline_scheduler.py [hunk 8]
```diff
@@ -85,10 +85,7 @@ def _load_accum_batch(self, data: Any, label: Any):
         self._grad_accum_offset += self._bsz_stride
 
         if self.data_process_func:
-            _data["input_ids"] = self.data_process_func(_data["input_ids"], _data["cu_seqlens"])
-            _label = self.data_process_func(_label, _data["cu_seqlens"], padding_v=-100)
-            _data.pop("cu_seqlens")
-            _data.pop("indexes")
+            _data, _label = self.data_process_func(_data, _label)
```
Comment (huangting4201) [inline internlm/core/scheduler/no_pipeline_scheduler.py:89]: 这块要同步下develop分支的bug修复

File: internlm/core/scheduler/pipeline_scheduler.py
Comment (huangting4201) [inline internlm/core/scheduler/pipeline_scheduler.py:1]: 这个文件要同步下develop分支的改动

File: internlm/data/utils.py [hunk 9]
```diff
@@ -24,34 +26,50 @@ def get_dataset_type_id(dataset_type_ids_map, path):
...
 
     for i in range(bsz):
-        output = torch.empty(num_seq, seq_len_, device=input_ids.device, dtype=dtype_).fill_(padding_v)
+        output = torch.empty(num_seq, seq_len_, device=data.device, dtype=dtype_).fill_(padding_v)
         cu_seqlens_slice = cu_seqlens[i]
         for j in range(num_seq):
             length = cu_seqlens_slice[j + 1] - cu_seqlens_slice[j]
-            output[j, 0:length] = input_ids[i, cu_seqlens_slice[j] : cu_seqlens_slice[j + 1]]
+            output[j, 0:length] = data[i, cu_seqlens_slice[j] : cu_seqlens_slice[j + 1]]
         outputs[i] = output
```
Comment (huangting4201) [inline internlm/data/utils.py:44]: 这个函数要同步下develop分支的改动

File: internlm/initialize/initialize_trainer.py [hunk 10]
```diff
@@ -79,10 +79,9 @@ def initialize_trainer(
...
     # initialize scheduler for trainer
     scheduler = None
-    if gpc.config.data.use_packed_dataset:
-        data_fn = None
-    else:
-        data_fn = unpack_data
+
+    data_fn = packed_data_normalizer if gpc.config.data.use_packed_dataset else unpack_data
+
     if gpc.is_using_parallel_mode(ParallelMode.PIPELINE):
```
Comment (huangting4201) [inline internlm/initialize/initialize_trainer.py:85]: 这里需同步develop分支的改动

File: internlm/initialize/launch.py
Comment (huangting4201) [inline internlm/initialize/launch.py:1]: 这个文件需要同步develop分支

File: internlm/core/parallel/shard.py [hunk 11]
```diff
@@ -0,0 +1,120 @@
...
+
+
+def get_parallel_strategies_split_mode(linear_name: str) -> str:
+    # TODO: 是否存在一种比根据名字来判断的方式？
+    tp_mode = gpc.config.parallel.tensor.mode
+
+    if linear_name in ("head", "output"):
+        return "head"
+    elif linear_name in ("wqkv", "wq", "wk", "wv", "wkv", "w1", "w3", "w13"):
+        return "column"
```
Comment (huangting4201) [inline internlm/core/parallel/shard.py:38]: 这里要加上一个Wqkv的选项

File: internlm/model/modules/embedding.py [hunk 12]
```diff
@@ -332,54 +136,62 @@ def _update_cos_sin_cache(self, x, indexes):
...
-        if kwargs.get("inference_params", None) is not None:
-            return self._eval_forward(qkv, seqlen_offset=kwargs.get("inference_params", None).sequence_len_offset)
+    def _get_slice(self, tensor: torch.Tensor, offsets: Union[int, torch.Tensor] = 0):
+        if isinstance(offsets, int):
+            return tensor[offsets:]
         else:
-            return self._eval_forward(qkv)
+            return tensor[offsets]
+
+    def _covert_padding(
```
Comment (huangting4201) [inline internlm/model/modules/embedding.py]: 这个应该是convert？

File: internlm/model/modules/linear.py [hunk 13]
```diff
@@ -0,0 +1,600 @@
...
+        return output if not return_residual else (output, x)
+
+    @staticmethod
+    @custom_bwd
+    def backward(ctx, grad_output, *args):
+        communicator: TPCommunicator = ctx.communicator
+
+        # parallel strategy-specific communication callback 3.
+        # see more details in the communicator for different parallel strategies.
+        grad_output, _ = communicator.grad_output_hook(grad_output, async_op=False)
```
Comment (huangting4201) [inline internlm/model/modules/linear.py:94]: 这个hook好像不需要？

File: internlm/core/parallel/comm/isp.py (hunk 2 above)
Comment (mwiacx) [reply internlm/core/parallel/comm/isp.py:694]: 我这里重构了不需要，你看下参数分发那里

File: internlm/core/parallel/comm/isp.py (hunk 3 above)
Comment (mwiacx) [reply internlm/core/parallel/comm/isp.py:722]: 同上

File: internlm/core/parallel/comm/isp.py (hunk 4 above)
Comment (mwiacx) [reply internlm/core/parallel/comm/isp.py:755]: 同上

File: internlm/core/parallel/shard.py (hunk 11 above)
Comment (mwiacx) [reply internlm/core/parallel/shard.py:38]: 入口有lower，不需要

Comment (mwiacx) [reply internlm/core/parallel/shard.py:38]: 他的caller，有lower

File: internlm/core/parallel/comm/isp.py (hunk 1 above)
Comment (mwiacx) [reply internlm/core/parallel/comm/isp.py]: 注释现在统一是没管的，等精度测试bug，调完了会统一整理

File: internlm/core/parallel/comm/utils.py (hunk 5 above)
Comment (mwiacx) [reply internlm/core/parallel/comm/utils.py:58]: 不是有reduce_dim参数吗？

Comment (mwiacx) [reply internlm/core/parallel/comm/utils.py:58]: 你刷新下？

File: internlm/core/parallel/comm/utils.py (hunk 6 above)
Comment (mwiacx) [reply internlm/core/parallel/comm/utils.py:226]: 你刷新下？

Comment (mwiacx) [reply internlm/core/parallel/comm/utils.py:226]: 一模一样的

File: internlm/core/parallel/shard.py (hunk 7 above)
Comment (mwiacx) [reply internlm/core/parallel/shard.py:44]:
```
        err_msg = (
            f"Parallel strategies for linear is unsupported, which is named as {name}.\n"
            + "Consider use manual_select_class parameter to select a linear class manually."
        )

        raise ValueError(err_msg)
```
caller有处理raise error

File: internlm/core/parallel/comm/isp.py (hunk 2 above)
Comment (mwiacx) [reply internlm/core/parallel/comm/isp.py:694]: 看下`params_dispatch_with_condition`和`check_attention_argument`

File: internlm/model/modules/embedding.py (hunk 12 above)
Comment (mwiacx) [reply internlm/model/modules/embedding.py]: ok

File: internlm/model/modules/linear.py (hunk 13 above)
Comment (mwiacx) [reply internlm/model/modules/linear.py:94]: `SequenceParallelCommunicator` 里面：
```
    def grad_output_hook(
        self, grad_output: torch.Tensor, async_op: bool = False
    ) -> Tuple[torch.Tensor, AsyncCommHandle]:
        """
        all gather grad_output only for row parallel linear when backward.
        """
        if dist.get_world_size(self._process_group) <= 1 or self._role == LinearRole.COLUMN:
            return grad_output, DUMMY_HANDLE_CONST

        return all_gather_raw(grad_output, process_group=self._process_group, async_op=async_op, gather_dim=_GATHER_DIM)
```

File: internlm/core/parallel/comm/isp.py (hunk 2 above)
Comment (huangting4201) [reply internlm/core/parallel/comm/isp.py:694]: 但是local_attn的参数列表是不确定的，比如说NPU的attn参数列表是这样的
```python
def forward(
        self,
        qkv=None,
        q=None,
        k=None,
        v=None,
        kv=None,
        cu_seqlens_q=None,  # pylint: disable=W0613
        cu_seqlens_k=None,  # pylint: disable=W0613
        max_seqlen_q=None,  # pylint: disable=W0613
        max_seqlen_k=None,  # pylint: disable=W0613
        dropout_p=0.0,
        softmax_scale=None,
        causal=False,
        window_size=(-1, -1),  # pylint: disable=W0613
        alibi_slopes=None,  # pylint: disable=W0613
        deterministic=False,
        return_attn_probs=False,  # pylint: disable=W0613
        attention_mask=None,
    ):
```

Comment (huangting4201) [reply internlm/core/parallel/comm/isp.py:694]: 看了下attention.py文件，了解了

Comment (mwiacx) [reply internlm/core/parallel/comm/isp.py:694]: 和 ops/attetion.py

File: internlm/model/ops/rotay_emb.py [hunk 14]
```diff
@@ -0,0 +1,156 @@
+"""
```
Comment (huangting4201) [inline internlm/model/ops/rotay_emb.py]: 这个文件名要更新一下

File: internlm/model/modules/mlp.py [hunk 15]
```diff
@@ -1,262 +1,113 @@
...
 
-def get_mlp_cls(tp_mode: str):
-    if tp_mode in ["mtp", "fsp"]:
-        mlp_cls = FeedForward
-    elif tp_mode == "msp":
-        mlp_cls = MegatronFeedForward
-    else:
-        mlp_cls = ISPFeedForward
-    return mlp_cls
+def new_fead_forward(
```
Comment (huangting4201) [inline internlm/model/modules/mlp.py]: 这里应该是feed

File: internlm/model/ops/rotay_emb.py (hunk 14 above)
Comment (mwiacx) [reply internlm/model/ops/rotay_emb.py]: ok

File: internlm/data/utils.py [hunk 16]
```diff
@@ -24,34 +26,50 @@ def get_dataset_type_id(dataset_type_ids_map, path):
...
+    # or let the dataset handle it? Currently inclined towards the latter.
+    assert data["input_ids"].shape[0] == 1, "data should be packed with batch size 1"
+
+    data["indexes"] = data["indexes"][0]
+    data["cu_seqlens"] = data["cu_seqlens"][0].squeeze(0)
+    data["max_seqlen"] = (data["cu_seqlens"][1:] - data["cu_seqlens"][:-1]).max().item()
+
+    # Move to parallel package for standardization
+    if gpc.config.parallel.sequence_parallel and gpc.config.parallel["tensor"].get("mode", "mtp") == "isp":
+        data["indexes"] = _split(data["indexes"], ParallelMode.TENSOR, dim=0)
```
Comment (huangting4201) [inline internlm/data/utils.py:74]: 如果我们在load micro batch时就把indexes切好了，那么backward时就不需要gather了对吧

Comment (mwiacx) [reply internlm/data/utils.py:74]: 是的，indexes只有rotary_emb用，主要是用于forward求sin,cos矩阵，backward会用ctx.sin, ctx.cos，所以没有必要在backward处理indexes

File: internlm/model/modules/mlp.py (hunk 15 above)
Comment (mwiacx) [reply internlm/model/modules/mlp.py]: ok

File: internlm/core/parallel/comm/isp.py [hunk 17]
```diff
@@ -30,6 +31,34 @@
...
+    def weight_hook(self, tensor: torch.Tensor, async_op: bool = False, **kwargs) -> torch.Tensor:
+        """
+        communiction for weight when forward/backward.
+        """
+        pass
+
+    @abstractmethod
+    def grad_hook(self, tensor: torch.Tensor, async_op: bool = False, **kwargs) -> Tuple[torch.Tensor, AsyncCommHandle]:
+        """
+        communiction for grad when backward.
```
Comment (huangting4201) [inline internlm/core/parallel/comm/isp.py]: 43行communictor，50行57行communiction拼写错了

Comment (mwiacx) [reply internlm/core/parallel/comm/isp.py]: done

File: internlm/train/pipeline.py [hunk 18]
```diff
@@ -325,10 +336,26 @@ def initialize_parallel_communicator(model: Union[nn.Module, nn.ModuleList]):
...
+                process_group=gpc.get_group(ParallelMode.TENSOR), role=LinearRole.ROW
+            )
+            for moe in _submodule_filter(model, MoE):
+                # 1. the linear in MoE degrades the parallel communication pattern from sp to tp
+                for column_linear in _submodule_filter(moe, ColumnParallelLinear):
+                    column_linear.register_communicator(_column_communicator)
+                for row_linear in _submodule_filter(moe, RowParallelLinear):
+                    row_linear.register_communicator(_row_communicator)
+                # 2. register MoESequenceParallelCommunicator for MoE layer
+                MoESequenceParallelCommunicator(ParallelMode.TENSOR).register_module_hook(moe)
```
Comment (huangting4201) [inline internlm/train/pipeline.py:362]: 这个应该只在sequence parallel为True的时候才注册？

Comment (huangting4201) [reply internlm/train/pipeline.py:362]: 外面包了一层if判断，OK的

**Review by sunpengsdu — APPROVED** [review]
